Add draft governance ZIPs

daira opened the pull request.

daira committed:
Add draft governance ZIPs.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

daira committed:
Updates to draft-ecc-community-and-coinholder based on feedback from
@nuttycom and Josh.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+* There is a well-defined, publically agreed, process for evaluation and community feedback on grant proposals.
+* Funds are held in a multisig resistant to compromise of some of the parties' keys, up to a given threshold.
+* No single party's non-cooperation or loss of keys is able to cause any Protocol-defined Ecosystem Funding to be locked up unrecoverably.
+* During the period of this proposal, the block rewards must be distributed as described in the [Abstract] above.
```
The word "distributed" in this sentence seems overloaded to me. Does this refer to the 8%/12% split, or the sentence that follows (which doesn't provide enough clarity to qualify as a requirement)?

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+* Funds are held in a multisig resistant to compromise of some of the parties' keys, up to a given threshold.
+* No single party's non-cooperation or loss of keys is able to cause any Protocol-defined Ecosystem Funding to be locked up unrecoverably.
+* During the period of this proposal, the block rewards must be distributed as described in the [Abstract] above.
+* The funds from the 8% and 12% funding streams will be usable immediately.
```
This requirement precludes the use of a buffering approach. Is that intended?

daira replied:
This wasn't true of the 12% stream (and in Option 1 there isn't any such stream). So "and 12%" has been deleted and this comment no longer applies. The 8% will continue to be provided via a funding stream without buffering.

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+* No single party's non-cooperation or loss of keys is able to cause any Protocol-defined Ecosystem Funding to be locked up unrecoverably.
+* During the period of this proposal, the block rewards must be distributed as described in the [Abstract] above.
+* The funds from the 8% and 12% funding streams will be usable immediately.
+* The funds in the Deferred Dev Fund Lockbox will be usable immediately.
```
"immediately" is not well-defined. At best, this should be "as of the activation height of the protocol change specified by this ZIP".

daira replied:
Changed to "usable immediately on activation of this ZIP".

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+* During the period of this proposal, the block rewards must be distributed as described in the [Abstract] above.
+* The funds from the 8% and 12% funding streams will be usable immediately.
+* The funds in the Deferred Dev Fund Lockbox will be usable immediately.
+* Any use of Deferred Dev Fund Lockbox funds is consistent with the purpose specified in [^zip-1015-lockbox] of "funding grants to ecosystem participants".
```
```suggestion
* Any use of Deferred Dev Fund Lockbox funds must be consistent with the purpose specified in [^zip-1015-lockbox] of "funding grants to ecosystem participants".
```

daira replied:
The other requirements use "is", "will be", etc. They should all be changed or none of them changed. I think it reads better as-is.

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+
+The Key-Holder Organizations would be bound by a legal agreement to only use funds held in the Coin-holder-controlled Fund according to the specifications in this ZIP, expressed in suitable legal language. In particular, all requirements on the use of Deferred Dev Fund Lockbox funds in ZIP 1015 [^zip-1015-lockbox] MUST be followed for the funds obtained from the one-time disbursal.
+
+The Financial Privacy Foundation (FPF) would administer the Coin-holder-controlled Fund based on decisions taken by coin-holder voting, following a model decided by the community and specified in another ZIP [TBD].
```
Note from Josh: 

>  FPF will not be involved. Some org (TBD) will need to be designated to conduct polling.

daira replied:
This now reads "The Key-Holder Organizations would collectively administer the Coinholder-Controlled Fund based on decisions taken by coinholder voting, ..."

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+
+### Disbursement process
+
+1. Anyone can submit a grant application via a process agreed between the FPF, the Zcash Foundation, and the Electric Coin Company.
```
Adjust to remove the FPF; probably this should be:
```suggestion
1. Anyone can submit a grant application via a process agreed upon by the Key-Holder Organizations.
```

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+1. Anyone can submit a grant application via a process agreed between the FPF, the Zcash Foundation, and the Electric Coin Company.
+2. Grant applications with total below USD 500,000 (or equivalent in another currency) would be directed to ZCG.
+3. For grant applications above this threshold, a 30-day community review and feedback period would start.
+4. If the grant applications should proceed to a coin-holder vote according to the agreed process, then coin-holders would be asked to vote on it.
```
```suggestion
4. If a grant application proceeds to a coin-holder vote according to the agreed process, then coin-holders will be asked to vote on it.
```
This doesn't specify a timeline; does the vote take place at the end of the 30 day window? As written, this looks like it allows a vote to be deferred indefinitely.

nuttycom replied:
Also: at present this doesn't define a coin-holder voting mechanism. Using Hanh's coin voting mechanism restricts votes to funds held in the Orchard pool; if that's intended, then it should be explicitly specified; however, that might be construed to contradict https://github.com/zcash/zips/pull/989/files#diff-8528553152727684ca93a225882ceed8e57a309ae833f5f517f0acbf573cd61aR129 which says that organizations cannot be restricted from voting their coins; if some barrier (such as lack of server-grade HSM support for Orchard) prevents an organization from safely voting their coins, is that a problem?

daira replied:
> This doesn't specify a timeline; does the vote take place at the end of the 30 day window?

As currently written, the "process agreed upon by the Key-Holder Organizations" decides that. So it's not ambiguous, but we might want to tighten it up.

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+2. Grant applications with total below USD 500,000 (or equivalent in another currency) would be directed to ZCG.
+3. For grant applications above this threshold, a 30-day community review and feedback period would start.
+4. If the grant applications should proceed to a coin-holder vote according to the agreed process, then coin-holders would be asked to vote on it.
+5. If the vote passes, then as payments are scheduled for the grant, then the FPF, the Zcash Foundation, and the Electric Coin Company SHOULD sign the corresponding transactions for disbursement from the Coin-holder-controlled Fund.
```
```suggestion
5. In order to ensure that subversion of the Coin-holder voting mechanism cannot result in draining of the Coin-holder-controlled Fund, and to ensure there is no disbursement of funds in a manner that is antithetical to the values of the Zcash community, the Key-holder Organizations MAY veto coin-holder approval via public announcement within 10 days of the tabulation of the coin-holder poll. If such a veto is exercised, the funds specified in the vetoed grant will remain at rest. The Key-holder Organizations MAY NOT veto coinholder rejection of a proposal.
6. If the coinholder vote approves the grant and no veto is exercised, then the Key-holder Organizations SHOULD sign transactions for disbursement from the Coin-holder-controlled Fund according to the schedule proposed by the grant. 
```

daira replied:
@str4d, @nuttycom and I discussed cases where there is disagreement between the Key-Holder Organizations so that they cannot come to a collective decision on a veto. In that case, I suggested that it should be sufficient for at least two of the Key-Holder Organizations to agree on a veto. (A threshold of one presents too much risk of a single org becoming an obstruction; a threshold of three is too high since this is a veto mechanism, not a voting mechanism.)

daira replied:
The current wording reflects that there are two criteria for a veto: legal or reporting requirements which have a threshold of 1, and other objections which have a threshold of 2.

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+
+For coin holder votes, a minimum of 420,000 ZEC (2% of the total eventual supply) MUST be voted, with a simple majority cast in favor, for a grant proposal to be approved. Upon approval, the grants would be paid to the recipient per the terms of the grant proposal.
+
+Coin holders would not be obligated to fund grants and MAY leave the funds at rest.
```
This sentence seems redundant to me, but if it's kept in I think it should be clarified:
```suggestion
Coin holders are not obligated to fund any grants and MAY leave the funds at rest.
```

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+
+Coin holders would not be obligated to fund grants and MAY leave the funds at rest.
+
+No organizations or individuals would be restricted from voting their coins.
```
This seems like it belongs in the `Requirements` section. It's not really specification.
```suggestion
No organizations or individuals may be restricted from voting their coins.
```

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+
+### One-time lockbox disbursal
+
+The coinbase transaction of the activation block of this ZIP MUST include an additional output to a 3-of-5 P2SH multisig with keys held by the Zcash Foundation, the Electric Coin Company, Shielded Labs, and two other organizations yet to be decided ("Key-holder Organizations").
```
nit: I think that `Key-Holder`, `Coin-Holder`, and `Coin-Holder Controlled` should have all words capitalized; to capitalize just the first word of the hyphenated phrase looks wrong to me.
```suggestion
The coinbase transaction of the activation block of this ZIP MUST include an additional output to a 3-of-5 P2SH multisig with keys held by the Zcash Foundation, the Electric Coin Company, Shielded Labs, and two other organizations yet to be decided ("Key-Holder Organizations").
```

If accepted, apply this suggestion throughout.

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+
+1. Anyone can submit a grant application via a process agreed between the FPF, the Zcash Foundation, and the Electric Coin Company.
+2. Grant applications with total below USD 500,000 (or equivalent in another currency) would be directed to ZCG.
+3. For grant applications above this threshold, a 30-day community review and feedback period would start.
```
```suggestion
3. For grant applications above this threshold, a 30-day community review and feedback period will start.
```

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+
+For the avoidance of doubt, this includes the following:
+
+* Decisions on use of the Coin-holder-controlled Fund taken by coin-holder vote require no approval from the FPF Board, but are subject to veto if FPF judges them to violate Cayman law or the FPF's reporting requirements and other (current or future) obligations under the Cayman Islands' Companies Act (2023 Revision) and Foundation Companies Law, 2017.
+* FPF SHALL be contractually required to recognize the Coin-holder-controlled Fund as a Restricted Fund donation under the above constraints, and keep separate accounting of its balance and usage under its Transparency and Accountability obligations defined in ZIP 1015 [^zip-1015-transparency-and-accountability].
```
If the FPF is not involved (as mentioned by Josh above) do these clauses need to be retained, or should they be modified to refer to the Key-Holder Organizations?

daira replied:
They don't need to be retained given the section on Veto process.

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,188 @@
+* Decisions on use of the Coin-holder-controlled Fund taken by coin-holder vote require no approval from the FPF Board, but are subject to veto if FPF judges them to violate Cayman law or the FPF's reporting requirements and other (current or future) obligations under the Cayman Islands' Companies Act (2023 Revision) and Foundation Companies Law, 2017.
+* FPF SHALL be contractually required to recognize the Coin-holder-controlled Fund as a Restricted Fund donation under the above constraints, and keep separate accounting of its balance and usage under its Transparency and Accountability obligations defined in ZIP 1015 [^zip-1015-transparency-and-accountability].
+
+Coin-holders SHOULD take account of the same factors considered by the ZCG Committee (described in points 2, 3, and 4 of [^zip-1015-zcg]) in deciding whether to fund a grant. If any contentious issue arises in connection with a grant (such as milestones not being met), or periodically for grants of indefinite duration, FPF SHOULD hold additional coin-holder votes to determine whether funding shou ...
```
Does this also need to be modified to refer to the Key-Holder Organizations instead of the FPF?
```suggestion
Coin-holders SHOULD take account of the same factors considered by the ZCG Committee (described in points 2, 3, and 4 of [^zip-1015-zcg]) in deciding whether to fund a grant. If any contentious issue arises in connection with a grant (such as milestones not being met), or periodically for grants of indefinite duration, the Key-Holder Organizations SHOULD administer additional coin-holder votes to determine whether funding should continue.
```

nuttycom reviewed (changes requested):
Reviewed draft-ecc-community-and-coinholder.md

nuttycom commented on zips/draft-ecc-zbloc.md:
```diff
@@ -0,0 +1,268 @@
+* The voting mechanism used for decision-making is similarly resistant to compromise of some of the parties' keys, up to a given threshold.
+* No single party's non-cooperation or loss of keys is able to cause any Protocol-defined Ecosystem Funding to be locked up unrecoverably.
+* 80% of the block rewards are given to miners, and 20% to a Zcash Community Fund seeded from the Deferred Dev Fund Lockbox.
+* The funds from the 20% funding stream will be usable immediately.
```
"immediately" is not well-defined; same issue as https://github.com/zcash/zips/pull/989/files#r2003460116

nuttycom commented on zips/draft-ecc-zbloc.md:
```diff
@@ -0,0 +1,268 @@
+* No single party's non-cooperation or loss of keys is able to cause any Protocol-defined Ecosystem Funding to be locked up unrecoverably.
+* 80% of the block rewards are given to miners, and 20% to a Zcash Community Fund seeded from the Deferred Dev Fund Lockbox.
+* The funds from the 20% funding stream will be usable immediately.
+* The funds in the Deferred Dev Fund Lockbox will be usable once a mechanism is available for disbursing them.
```
Should the disbursement mechanism options be factored out into a separate ZIP that both the governance ZIPs refer to? In both proposals, the model is effectively "grant-based funding, where development funds are controlled by a multisig" with different options for how and when funds may be disbursed to grantees.

jswihart replied:
Will someone take responsibility to draft this?

daira replied:
@nuttycom agreed to do this.

nuttycom commented on zips/draft-ecc-zbloc.md:
```diff
@@ -0,0 +1,268 @@
+
+Any changes to the ZCG or its governance, such as its expansion to include more members, may be desired but are specifically outside this proposal’s scope.
+
+The full implementation of this proposal would require the design and implementation of a lockbox distribution mechanism, likely using multisig. This mechanism would be similar to that needed for an alternative proposal, the Community and Coin-holder Funding Model [^draft-ecc-community-and-coinholder]. However, this proposal does not require that a lockbox distribution mechanism be implemented at ...
```
This gets to my comment above: any of the ecc-community-and-coinholder options would work for this proposal as well.

nuttycom commented on zips/draft-ecc-zbloc.md:
```diff
@@ -0,0 +1,268 @@
+
+Each Constituency initially receives three Voting Units. This MAY be changed by a Governance Proposal.
+
+Constituencies other than the Coin-holder Constituency above are called Organizational Constituencies. The Organizational Constituencies will together elect a representative responsible for establishing a clear process and mechanism for gathering coin-holder sentiment, which will then be used as the decision-making process of the Coin-holder Constituency.
```
One possibility that this doesn't take into account is that an organic coin-holder constituency may choose to self-organize; the way that this is worded makes it sounds like the Organization Constituencies will choose the decision-making process instead of just the mechanism for determining coin-holder sentiment.
```suggestion
Constituencies other than the Coin-holder Constituency above are called Organizational Constituencies. The Organizational Constituencies will together elect a representative responsible for establishing a clear process and mechanism for gathering coin-holder sentiment, which will then be used to determine the will of the Coin-holder Constituency.
```

nuttycom commented on zips/draft-ecc-zbloc.md:
```diff
@@ -0,0 +1,268 @@
+
+### Approval Thresholds
+
+The Approval Threshold is the absolute majority, as a proportion of the total number of Eligible Voting Units, required for a Proposal to be Approved. As stated earlier, Eligible Voting Units do not include those of Constituencies that have a conflict of interest for this Proposal. However, they do include the Voting Units of Constituencies that have no conflict of interest but choose to abstain.
```
We need something better than this: we need a system that makes it easy to choose between multiple possible alternatives. For example, consider the situation wherein ECC, ZF, Shielded Labs, and the Zingoistas all submit grant proposals for the same work; in this circumstance, is the grantee selected by determining which of the competing proposals has the highest number of approvals? What happens in the case of ties? 

In addition, timing constraints based upon when a grant proposal is made can introduce problems - for example, if a vote is scheduled for a grant, and then a few days before the vote a competing proposal is introduced, what happens? If a competing, but better proposal is introduced after the grant has been approved, what happens?

This comment also applies to the coinholder-and-community ZIP.

nuttycom replied:
Note that if additional proposals are made beyond these first two, we're *already* having to face this problem. It will be very common - it will probably arise for literally every proposal.

jswihart replied:
It would be approved when the signing threshold is reached. At that point, it becomes a commitment that would only be overturned if a new threshold signals that it should be.

nuttycom replied:
@jswihart the problem that I'm pointing out is that if there are multiple competing proposals, voting on each one individually doesn't work because you might end up in a situation where both grants are approved, but the community only wants to spend money on one of them and there's no way to signal that using the currently-described voting mechanism.

daira replied:
I was originally thinking that negotiation about alternatives could be done off-chain, and then the on-chain vote would just be a ratification. If we want to put more of the sausage-making on-chain then we can, but I'm not sure that is a good idea, because choosing between multiple alternatives [fundamentally](https://en.wikipedia.org/wiki/Arrow%27s_impossibility_theorem) requires a more complicated voting mechanism.

nuttycom commented on zips/draft-ecc-zbloc.md:
```diff
@@ -0,0 +1,268 @@
+
+The Approval Threshold is the absolute majority, as a proportion of the total number of Eligible Voting Units, required for a Proposal to be Approved. As stated earlier, Eligible Voting Units do not include those of Constituencies that have a conflict of interest for this Proposal. However, they do include the Voting Units of Constituencies that have no conflict of interest but choose to abstain.
+
+By default, a Governance Proposal requires a three-fifths Approval Threshold, and MUST have a Deadline at least two weeks from when the Proposal is published on-chain.
```
This seems out-of-order; we don't yet have any information about what it means to publish a Proposal on-chain, or who is responsible for that process.

daira replied:
It was hard to find a sensible order that didn't involve forward-references. There might be a better order, but in any case remaining forward-references should be explicit links; feel free to fix that.

nuttycom commented on zips/draft-ecc-zbloc.md:
```diff
@@ -0,0 +1,268 @@
+
+By default, a Governance Proposal requires a three-fifths Approval Threshold, and MUST have a Deadline at least two weeks from when the Proposal is published on-chain.
+
+All Consituencies MUST be consulted in order to determine whether they consider a draft Proposal to be "controversial", before it published on-chain. If *any* Constituency decides that a Proposal should be treated as controversial, it is then subject to a higher Approval Threshold of three quarters, and a Deadline at least four weeks from when the Proposal is published on-chain. (In the case of t ...
```
This strikes me as a can of worms. This means that an initial coinholder vote must be held to determine whether a proposal is "controversial" for every proposal.

daira replied:
I see your point. The intent was that a coinholder vote would only need to be done to confirm that a proposal is "controversial" if a coalition of people who plausibly claim to collectively have significant holdings are arguing that it is.

nuttycom commented on zips/draft-ecc-zbloc.md:
```diff
@@ -0,0 +1,268 @@
+* Increasing or decreasing the voting power of zBloc Constituencies by adjusting how many Voting Units they hold.
+* Increasing or decreasing the number of Funding Pools, the rules that govern them, and the amounts allocated.
+
+In the case of Governance Proposals that affect consensus, final ratification occurs when node operators adopt the software implementing the changes.
+
+Note: Nothing forces developers of Zcash consensus node software to follow this specification. The aim of a process specification like this one is only to establish a social consensus. It fundamentally cannot affect the autonomy of developers of Zcash consensus node software to publish (or not) the software they want to publish, or the autonomy of node operators to run (or not) the software they  ...
```
These facts are also true for the `community-and-coinholder` proposal (or, indeed, any proposal.) They should be copied there.

daira replied:
True, but community-and-coinholder is only attempting to handle funding proposals, and so it's obvious in that case that it's not making any demands of node developers.

nuttycom commented on zips/draft-ecc-zbloc.md:
```diff
@@ -0,0 +1,268 @@
+* Substantive changes to the ZIP Process defined in ZIP 0 [^zip-0000].
+* Substantive changes to this ZIP defining the zBloc.
+* Changes to the set of zBloc Constituencies and/or the distribution of Voting Units among those Constituencies.
+* Changes to the policy a Constituency uses for its decision-making.
```
This last one should only apply to the coinholder constituency; it should not require a Governance Proposal for ECC to change its internal decision-making mechanism.

daira replied:
Oh, I misinterpreted Josh's intent here then (given his thumbs-up).

nuttycom commented on zips/draft-ecc-zbloc.md:
```diff
@@ -0,0 +1,268 @@
+
+### Exceptional decisions to remediate security vulnerabilities
+
+The governance processes specified in this ZIP are not intended to prevent consensus node implementors from making consensus changes that they reasonably believe are critical to maintain the security of the Zcash Mainnet network and/or Zcash users. Such changes might need to be developed and released according to a controlled disclosure process that precludes public voting.
```
This is also useful to copy to the community-and-coinholder proposal, or to factor out into a shared preamble.

daira replied:
Again, community-and-coinholder only handles grants, not any other aspect of governance.

nuttycom commented on zips/draft-ecc-zbloc.md:
Raising again that I don't like the funding mechanism to be coupled to the governance mechanism. Also, is the FPF intended to be included here?

What is the rationale behind the keyholders being different from the constituencies? I can see where the coinholder vote would need its disbursement power to be split among the other constituencies, but otherwise it's not obvious why we should have separate keyholders. Add a rationale if this is actually the chosen design.

nuttycom reviewed (dismissed):
Review complete. Addressing the question of how to handle competing proposals in particular is blocking.

daira committed:
Updates based on @str4d and @nuttycom's comments.

Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

jswihart commented on rendered/draft-ecc-community-and-coinholder.html:
```diff
@@ -30,6 +30,8 @@ <h1 id="terminology"><span class="section-heading">Terminology</span><span class
 
 <p>The terms &#8220;Zcash Community Grants&#8221; (or &#8220;ZCG&#8221;), &#8220;ZCG Committee&#8221;, &#8220;Financial Privacy Foundation&#8221; (or &#8220;FPF&#8221;), and &#8220;Deferred Dev Fund Lockbox&#8221; in this document are to be interpreted as defined in ZIP 1015 <a href="#fn:4" id="fnref:4" title="see footnote" class="footnote"><sup>4</sup></a>.</p>
 
```
Remove FPF reference.

jswihart commented on rendered/draft-ecc-community-and-coinholder.html:
```diff
@@ -71,42 +72,55 @@ <h1 id="requirements"><span class="section-heading">Requirements</span><span cla
 <h1 id="non-requirements"><span class="section-heading">Non-requirements</span><span class="section-anchor"> <a rel="bookmark" href="#non-requirements"><img width="24" height="24" class="section-anchor" src="assets/images/section-anchor.png" alt=""></a></span></h1>
 
 <p>Any changes to the ZCG or its governance, such as its expansion to include more members, may be desired but are specifically outside this proposal’s scope.</p>
 
```
Any changes to Zcash protocol governance, specifically what changes are made to consensus node software, are outside this proposal's scope.

jswihart commented on rendered/draft-ecc-community-and-coinholder.html:
```diff
@@ -71,42 +72,55 @@ <h1 id="requirements"><span class="section-heading">Requirements</span><span cla
+
+<h3 id="option1:extendthelockboxfundingstream"><span class="section-heading">Option 1: Extend the lockbox funding stream</span><span class="section-anchor"> <a rel="bookmark" href="#option1:extendthelockboxfundingstream"><img width="24" height="24" class="section-anchor" src="assets/images/section-anchor.png" alt=""></a></span></h3>
+
+<p>This option extends the <code>FS_DEFERRED</code> lockbox funding stream by a further 1,260,000 blocks (approximately 3 years), ending at the 3rd halving.</p>
```
I did not intend for this to be an option.

nuttycom replied:
This is the option that we discussed yesterday, that includes making the necessary changes for a specific lockbox disbursement feature to the V6 transaction format. We added a rationale discussing that that during our pairing yesterday; @daira where did that rationale go?

nuttycom replied:
Oh, I see: @jswihart your review didn't include the commit that @daira added after our pairing yesterday. See https://github.com/zcash/zips/pull/989/files#diff-8528553152727684ca93a225882ceed8e57a309ae833f5f517f0acbf573cd61aR103-R109

jswihart commented on rendered/draft-ecc-community-and-coinholder.html:
```diff
@@ -71,42 +72,55 @@ <h1 id="requirements"><span class="section-heading">Requirements</span><span cla
+
+<p>The Key-Holder Organizations would be bound by a legal agreement to only use funds held in the Coin-holder-controlled Fund according to the specifications in this ZIP, expressed in suitable legal language. In particular, all requirements on the use of Deferred Dev Fund Lockbox funds in ZIP 1015 <a href="#fn:8" title="see footnote" class="footnote"><sup>8</sup></a> MUST be followed for the fund ...
 
 <p>The Financial Privacy Foundation (FPF) would administer the Coin-holder-controlled Fund based on decisions taken by coin-holder voting, following a model decided by the community and specified in another ZIP [TBD].</p>
```
Remove the FPF.

Shielded Labs would be responsible for coin-holder polling. 

The process would be that properly submitted grants (via some TBD mechanism) would be subject to a 30 day review period by the community before coin holder polling begins. SL would be responsible for conducting the poll and communicating the results.

daira replied:
This was already fixed. It's fine, but may I ask how you used the GitHub interface to end up reviewing only one commit? The usual way is to comment under the "Files changed" tab (unless you specifically intend to comment on a previous state).

jswihart commented on rendered/draft-ecc-community-and-coinholder.html:
```diff
@@ -128,13 +142,17 @@ <h3 id="disbursementprocess"><span class="section-heading">Disbursement process<
 <ul>
 <li>Decisions on use of the Coin-holder-controlled Fund taken by coin-holder vote require no approval from the FPF Board, but are subject to veto if FPF judges them to violate Cayman law or the FPF&#8217;s reporting requirements and other (current or future) obligations under the Cayman Islands' Companies Act (2023 Revision) and Foundation Companies Law, 2017.</li>
-<li>FPF SHALL be contractually required to recognize the Coin-holder-controlled Fund as a Restricted Fund donation under the above constraints, and keep separate accounting of its balance and usage under its Transparency and Accountability obligations defined in ZIP 1015 <a href="#fn:12" id="fnref:12" title="see footnote" class="footnote"><sup>12</sup></a>.</li>
+<li>FPF SHALL be contractually required to recognize the Coin-holder-controlled Fund as a Restricted Fund donation under the above constraints, and keep separate accounting of its balance and usage under its Transparency and Accountability obligations defined in ZIP 1015 <a href="#fn:11" id="fnref:11" title="see footnote" class="footnote"><sup>11</sup></a>.</li>
```
FPF is only responsible for its existing obligations with ZCG, but not any new obligations.

daira replied:
This has been removed.

jswihart commented on rendered/draft-ecc-community-and-coinholder.html:
```diff
@@ -128,13 +142,17 @@ <h3 id="disbursementprocess"><span class="section-heading">Disbursement process<
+<li>FPF SHALL be contractually required to recognize the Coin-holder-controlled Fund as a Restricted Fund donation under the above constraints, and keep separate accounting of its balance and usage under its Transparency and Accountability obligations defined in ZIP 1015 <a href="#fn:11" id="fnref:11" title="see footnote" class="footnote"><sup>11</sup></a>.</li>
 </ul>
 
 <p>Coin-holders SHOULD take account of the same factors considered by the ZCG Committee (described in points 2, 3, and 4 of <a href="#fn:7" title="see footnote" class="footnote"><sup>7</sup></a>) in deciding whether to fund a grant. If any contentious issue arises in connection with a grant (such as milestones not being met), or periodically for grants of indefinite duration, FPF SHOULD hold addi ...
```
Change FPF to SL.

jswihart commented on zips/draft-ecc-onchain-accountable-voting.md:
```diff
@@ -0,0 +1,175 @@
+
+The contents of the file at the given URL MUST NOT be automatically downloaded unless that is requested by an explicit user action.
+
+TODO: decide whether we want to just assume that memo bundles are available.
```
I suggest not tightly coupling this ZIP with feature features.

nuttycom replied:
I agree, we don't need a memo size change for the protocol as currently described to work, so I think we can get rid of this TODO.
```suggestion
```

daira replied:
The reason for the TODO was that *if* we knew that memo bundles were available, then it would be much simpler to just put the whole proposal on-chain rather than having to use a hash to check what is downloaded from the URL. But it seems like the intent is for it to be able to activate before the upgrade with memo bundles, so yes I will remove this TODO.

jswihart commented on zips/draft-ecc-zbloc.md:
```suggestion
Thank you to Tatyana Peacemonger, Kris Nuttycombe, and Chris Tomeo for the reviews, feedback, questions, and suggestions needed to make this a strong proposal.
```

daira committed:
Address some comments from @jswihart's review.

Co-authored-by: Josh Swihart <josh@electriccoin.co>
Co-authored-by: Kris Nuttycombe <kris@nutty.land>
Signed-off-by: Daira-Emma Hopwood <daira@jacaranda.org>

nuttycom committed:
Factor out lockbox distribution and funding stream mechanisms into a separate ZIP draft.

daira commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -66,7 +65,6 @@ It would immediately increase coinholders’ voice, minimize governance confusio
 * The funds in the Deferred Dev Fund Lockbox will be usable immediately on activation of this ZIP.
 * Any use of Deferred Dev Fund Lockbox funds is consistent with the purpose specified in [^zip-1015-lockbox] of "funding grants to ecosystem participants".
 
-
```
It's not important (and makes no difference to rendering), but we generally leave two lines between top-level sections.

daira commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -86,49 +84,7 @@ This funding stream will start on expiry of the existing ``FS_FPF_ZCG`` funding
 
```
TODO: check whether this is correct for Testnet.

daira commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -86,49 +84,7 @@ This funding stream will start on expiry of the existing ``FS_FPF_ZCG`` funding
-Classic funding streams [^zip-0207] produce many small output values, due to only being able to direct funds from a single block's subsidy at a time. This creates operational burdens to utilizing the funds — in particular due to block and transaction sizes limiting how many outputs can be combined at once, which increases the number of required transactions and correspondingly the overall fee.
-
-The periodic lockbox disbursement mechanism can produce the same effective funding stream, but with aggregation performed for free: the output to the funding stream recipient is aggregated into larger outputs every $N$ blocks. In the specific case of Mechanism 2b, the recipient multisig address would receive around 40 outputs, instead of around 1,300,000.
+The pool of multisig-controlled funds described as the Zcash Development Fund by [^draft-ecc-lockbox-distribution] forms a new Coinholder-Controlled Fund.
```
Calling it the Zcash Development Fund is confusing because in ZIPs 1014 and 1015, that included the ZCG Slice.

nuttycom replied:
I don't know how to generalize this name, because in the community-and-coinholder proposal this fund receives 12% of the block reward, and in the ZBloc proposal it receives 20%.

nuttycom replied:
Figured it out.

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -144,7 +100,9 @@ The Key-Holder Organizations collectively administer the Coinholder-Controlled F
 5. If the vote passes, then as payments are scheduled for the grant, then (subject to the [Veto process]) the Key-Holder Organizations SHOULD sign the corresponding transactions for disbursement from the Coinholder-Controlled Fund.
 
-For coinholder votes, a minimum of 420,000 ZEC (2% of the total eventual supply) MUST be voted, with a simple majority cast in favor, for a grant proposal to be approved. Upon approval, the grants are paid to the recipient per the terms of the grant proposal.
+For coinholder votes, a minimum of 420,000 ZEC (2% of the total eventual supply) MUST be voted, with a simple majority cast in favor, for a grant proposal to be approved. Upon approval, the grants are paid to the recipient per the terms of the grant proposal. In the case that multiple grant proposals are submitted that are in competition with one another, a single winner will be selected by holdi ...
```
Good point; the Key-Holder Organizations (since they're running the coinholder votes) will need to make this determination.

nuttycom committed:
Ensure consistency between lockbox disbursement & governance proposals.

This also eliminates some confusing naming, and specifies parameters to
the lockbox disbursement proposal more precisely.

str4d commented:
I have not performed a close review, but I've checked that these drafts are sufficiently complete to be mergeable into the repo as drafts without assigned ZIP numbers (per the current ZIP process we've used for previous recent ZIP drafts, which is represented on the website but we still need to update ZIP 0 with). Further changes can be made as subsequent PRs.

Once the ZIP Authors notify the ZIP Editors that the drafts are ready for review, the ZIP Editors can consider the drafts in a subsequent ZIP Sync.

nuttycom committed:
zbloc: Linkify "Governance Proposal" references

daira commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,178 @@
+
+## Coinholder-Controlled Fund
+
+A pool of multisig-controlled funds, seeded from the existing contents of the ZIP 1015 Deferred Dev Fund Lockbox [^zip-1015-lockbox] and supplemented with a funding stream consising of 12% of the block subsidy for the same time period as the Zcash Community Grants stream, forms a new Coinholder-Controlled Fund. The mechanisms for the creation and management of this fund are described by the Defer ...
```
s/mainnet/Mainnet/

nuttycom replied:
Addressed in https://github.com/zcash/zips/pull/994

daira commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,178 @@
+ 
+### Requirements on use of the Coinholder-Controlled Fund
+
+The Key-Holder Organizations SHALL be bound by a legal agreement to only use funds held in the Coinholder-Controlled Fund according to the specifications in this ZIP, expressed in suitable legal language. In particular, all requirements on the use of Deferred Dev Fund Lockbox funds in ZIP 1015 [^zip-1015-lockbox] MUST be followed for Coinholder-Controlled Fund.
```
s/for /for the /

nuttycom replied:
Addressed in https://github.com/zcash/zips/pull/994

daira commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,178 @@
+
+For coinholder votes, a minimum of 420,000 ZEC (2% of the total eventual supply) MUST be voted, with a simple majority cast in favor, for a grant proposal to be approved. Upon approval, the grants are paid to the recipient per the terms of the grant proposal. In the case that multiple grant proposals are submitted that are in competition with one another, a single winner will be selected by holdi ...
+
+Each coinholder vote (including in cases where multiple grants are in competition) is independent, and coins used in one vote are also allowed to be used in another; this approach is necessary to avoid vote-splitting scenarios that can result in an option being selected that achieves only a plurality (not a majority) of coinholder support.
```
Is it clear enough to non-voting-system-nerds what this means, or do we need more explanation?

nuttycom replied:
As someone who's not in the set of non-voting-system-nerds, I can't judge!

daira commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,178 @@
+
+The provisions after the first paragraph of the section "Zcash Community Grants (ZCG)" also apply to the Key-Holder Organizations' administration of the Coinholder-Controlled Fund, with coinholder voting replacing the role of the ZCG Committee.
+
+Note: Nothing forces developers of Zcash consensus node software to implement any particular proposal. The aim of a process specification like this one is only to social consensus. It fundamentally cannot affect the autonomy of developers of Zcash consensus node software to publish (or not) the software they want to publish, or the autonomy of node operators to run (or not) the software they want ...
```
"only to indicate social consensus"? or "establish"?

daira commented on zips/draft-ecc-lockbox-disbursement.md:
```diff
@@ -0,0 +1,245 @@
+  funding streams defined in ZIP 1015 [^zip-1015-funding-streams].
+* A variety of different mechanisms that may be used for distributing funds
+  that accrue during the period of the extension, which may vary depending upon
+  which proposal this mechanism is used.
```
"the proposal that uses this mechanism".

nuttycom replied:
Addressed in https://github.com/zcash/zips/pull/994

daira commented on zips/draft-ecc-lockbox-disbursement.md:
```diff
@@ -0,0 +1,245 @@
+  keys, up to a given threshold.
+* No single party's non-cooperation or loss of keys is able to cause any
+  Protocol-defined Ecosystem Funding to be locked up unrecoverably.
+* The funds previously accrued the Deferred Dev Fund Lockbox as the activation
```
"accrued by" or "accrued to"

daira commented on zips/draft-ecc-lockbox-disbursement.md:
```diff
@@ -0,0 +1,245 @@
+
+The ``FS_DEFERRED`` lockbox funding stream is set to receive
+$\mathsf{stream\_value}\%$ of the block subsidy and is extended until block
+height $\mathsf{stream\_end\_height}$ Both of these parameters must must be
```
Missing ".".

nuttycom replied:
Addressed in https://github.com/zcash/zips/pull/994

daira commented on zips/draft-ecc-lockbox-disbursement.md:
```diff
@@ -0,0 +1,245 @@
+format changes because such changes are disruptive to the ecosystem. It is not
+necessary that protocol rules for disbursement actually be implemented until
+after the transaction format changes are live on the network. It is RECOMMENDED
+that any such transaction format changes be included in the upcoming V6
```
"v6" (consistent capitalization).

nuttycom replied:
Addressed in https://github.com/zcash/zips/pull/994

daira commented on zips/draft-ecc-lockbox-disbursement.md:
```diff
@@ -0,0 +1,245 @@
+network upgrade to recover from key loss or compromise, or to change the size
+of the signing set or the number of signatures required to reach threshold.
+
+#### Security implications for Option 2a
```
"for Mechanism 2a".

nuttycom replied:
Addressed in https://github.com/zcash/zips/pull/994

daira commented on zips/draft-ecc-lockbox-disbursement.md:
```diff
@@ -0,0 +1,245 @@
+#### Security implications for Option 2b
+
+Due to the aggregation of funds recommended by Option 2b, it is no longer
+necessary use scripts with elevated privileges to perform shielding and/or
```
"necessary to use"

nuttycom replied:
Addressed in https://github.com/zcash/zips/pull/994

daira commented on zips/draft-ecc-onchain-accountable-voting.md:
```diff
@@ -0,0 +1,173 @@
+
+# Specification
+
+A Proposal to be decided using On-chain Accountable Voting is represented as Proposal Transaction, which references a human-readable Description of what is to happen on its Approval. The Proposal Transaction MAY also lock up funds that are to be granted on Approval [TODO].
```
"as a Proposal Transaction"

nuttycom replied:
https://github.com/zcash/zips/pull/994

jswihart commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,178 @@
+    ZIP: XXX
+    Title: Community and Coinholder Funding Model
+    Owners: Josh Swihart <josh@electriccoin.co>
+    Credits: Daira-Emma Hopwood
+             Kris Nuttycombe
+             Jack Grigg
+             Tatyana Peacemonger
```
Add Alex Bornstein and Jason McGee to credits

nuttycom replied:
https://github.com/zcash/zips/pull/994

jswihart commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,178 @@
+### Disbursement process
+
+1. Anyone can submit a grant application via a process agreed upon by the Key-Holder Organizations.
+2. Grant applications with total below USD 500,000 (or equivalent in another currency) are directed to ZCG.
```
Remove " Grant applications with total below USD 500,000 (or equivalent in another currency) are directed to ZCG."

jswihart commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,178 @@
+
+* Is a 3-of-5 threshold appropriate? What should the other two Key-Holder Organizations be?
+* Is 420,000 ZEC the right threshold for coinholder votes?
+* Is USD 500,000 the right threshold for grants to be controlled by coinholder vote rather than ZCG?
```
Remove

nuttycom replied:
https://github.com/zcash/zips/pull/994

jswihart commented on zips/draft-ecc-lockbox-disbursement.md:
```diff
@@ -0,0 +1,245 @@
+The coinbase transaction of the activation block of this ZIP MUST include an
+additional output to a 3-of-5 P2SH multisig with keys held by the following
+"Key-Holder Organizations": Zcash Foundation, the Electric Coin Company,
+Shielded Labs, and two other organizations yet to be decided.
```
Change to 2 of 3 (ECC, SL, ZF) initially but open to adding more as more organizations are willing and able to participate.

nuttycom replied:
Since this is the one-time disbursement section, I'm not clear on what the edit should be to support the "open to adding more orgs" bit. We should probably just make it 2-of-3 for now and then, if other orgs step up, add them if appropriate?

nuttycom replied:
Addressed in https://github.com/zcash/zips/pull/994

jswihart commented on zips/draft-ecc-zbloc.md:
Remove FPF

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,178 @@
+
+# Open questions
+
+* Is a 3-of-5 threshold appropriate? What should the other two Key-Holder Organizations be?
```
https://github.com/zcash/zips/pull/994

nuttycom commented on zips/draft-ecc-community-and-coinholder.md:
```diff
@@ -0,0 +1,178 @@
+
+1. Anyone can submit a grant application via a process agreed upon by the Key-Holder Organizations.
+2. Grant applications with total below USD 500,000 (or equivalent in another currency) are directed to ZCG.
+3. For grant applications above this threshold, a 30-day community review and feedback period will start.
```
https://github.com/zcash/zips/pull/994

Justfunnyboy1 reviewed (commented):
Me

Justfunnyboy1 reviewed (commented):
Euper
